pep-0544.txt

```diff
@@ -0,0 +1,1006 @@
+
+  var: Type[Proto]
+  var = Proto    # Error
+  var = Concrete # OK
+  var().meth()   # OK
+
+Assigning an ABC or a protocol class to a variable is allowed if it is
+not explicitly typed, and such assignment creates a type alias.
+For normal (non-abstract) classes, the behavior of ``Type[]`` is
+not changed.
+
+
+``NewType()`` and type aliases
+------------------------------
+
+Protocols are essentially anonymous. To emphasize this point, static type
+checkers might refuse protocol classes inside ``NewType()`` to avoid an
+illusion that a distinct type is provided::
+
+  form typing import NewType , Protocol, Iterator
+
+  class Id(Protocol):
+      code: int
+      secrets: Iterator[bytes]
+
+  UserId = NewType('UserId', Id)  # Error, can't provide distinct type
+
+On the contrary, type aliases are fully supported, including generic type
+aliases::
+
+  from typing import TypeVar, Reversible, Iterable, Sized
+
+  T = TypeVar('T')
+  class SizedIterable(Iterable[T], Sized, Protocol):
+      pass
+  CompatReversible = Union[Reversible[T], SizedIterable[T]]
+
+
+.. _discussion:
+
+``@runtime`` decorator and narrowing types by ``isinstance()``
+--------------------------------------------------------------
+
+The default semantics is that ``isinstance()`` and ``issubclass()`` fail
+for protocol types. This is in the spirit of duck typing -- protocols
+basically would be used to model duck typing statically, not explicitly
+at runtime.
+
+However, it should be possible for protocol types to implement custom
+instance and class checks when this makes sense, similar to how ``Iterable``
+and other ABCs in ``collections.abc`` and ``typing`` already do it,
+but this is limited to non-generic and unsubscripted generic protocols
+(``Iterable`` is statically equivalent to ``Iterable[Any]`).
+The ``typing`` module will define a special ``@runtime`` class decorator
+that provides the same semantics for class and instance checks as for
+``collections.abc`` classes, essentially making them "runtime protocols"::
+
+  from typing import runtime, Protocol
+
+  @runtime
+  class Closeable(Protocol):
+      def close(self):
+          ...
+
+  assert isinstance(open('some/file'), Closeable)
+
+Static type checkers will understand ``isinstance(x, Proto)`` and
+``issubclass(C, Proto)`` for protocols defined with this decorator (as they
+already do for ``Iterable`` etc.). Static type checkers will narrow types
+after such checks by the type erased ``Proto`` (i.e. with all variables
+having type ``Any`` and all methods having type ``Callable[..., Any]``).
+Note that ``isinstance(x, Proto[int])`` etc. will always fail in agreement
+with PEP 484. Examples::
+
+  from typing import Iterable, Iterator, Sequence
+
+  def process(items: Iterable[int]) -> None:
+      if isinstance(items, Iterator):
+          # 'items' have type 'Iterator[int]' here
+      elif isinstance(items, Sequence[int]):
+          # Error! Can't use 'isinstance()' with subscripted protocols
```

Comment:
I believe that's not unique to protocols -- it's the same for other generic classes.